llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -851,7 +852,29 @@ TableFiles processOneModule(std::unique_ptr<Module> M, bool IsEsimd,
   bool IsLLVMUsedRemoved = false;
   if (GlobalVariable *GV = M->getGlobalVariable("llvm.used")) {
     assert(GV->user_empty() && "unexpected llvm.used users");
+    Constant *Initializer = GV->getInitializer();
+    GV->setInitializer(nullptr);
     GV->eraseFromParent();
+
+    // Destroy the initializer and all operands of it.
+    SmallVector<Constant *, 8> IOperands;
+    for (auto It = Initializer->op_begin(); It != Initializer->op_end(); It++) {
+      assert(isa<Constant>(*It) && "Unexpected initializer of llvm.used");
+      IOperands.push_back(dyn_cast<Constant>(*It));
+    }
+    assert(llvm::isSafeToDestroyConstant(Initializer) &&
+           "Cannot remove initializer of llvm.used global");
+    Initializer->destroyConstant();
+    for (auto It = IOperands.begin(); It != IOperands.end(); It++) {
+      assert(llvm::isSafeToDestroyConstant(*It) &&
+             "Cannot remove an element of initializer of llvm.used global");
+      auto F = dyn_cast<Function>((*It)->getOperand(0));
+      (*It)->destroyConstant();
+      // Remove unused kernel declarations to avoid LLVM IR check fails.
+      if (F->isDeclaration())
```

Comment:
May there be a function definition?

Reply:
F is either a kernel definition, in this case F needs to be processed in GenXSPIRVWriterAdaptor() and to do that it must not have users. Or F is a declaration, which needs removal to preserve IR correctness. Same F is not met as both declaration and definition in llvm.used global